pkg/scheduler/plugins/capacity/capacity.go

```diff
@@ -220,133 +402,387 @@ func (cp *capacityPlugin) OnSessionOpen(ssn *framework.Session) {
+	cp.queueOpts[queue.UID].parents = append(cp.queueOpts[parentInfo.UID].parents, parentInfo.UID)
+	return nil
+}
 
-		inqueue := r.LessEqualWithDimension(attr.realCapability, minReq)
-		klog.V(5).Infof("job %s inqueue %v", job.Name, inqueue)
-		if inqueue {
-			attr.inqueue.Add(job.DeductSchGatedResources(minReq))
-			return util.Permit
+func (cp *capacityPlugin) checkHierarchicalQueue(attr *queueAttr) error {
+	totalGuarantee := api.EmptyResource()
+	totalDeserved := api.EmptyResource()
+	for _, childAttr := range attr.children {
+		totalDeserved.Add(childAttr.deserved)
+		totalGuarantee.Add(childAttr.guarantee)
+		// Check if the parent queue's capability is less than the child queue's capability
+		if attr.capability != nil && childAttr.capability != nil && attr.capability.LessPartly(childAttr.capability, api.Zero) {
+			return fmt.Errorf("queue <%s> capability is less than its child queue <%s>", attr.name, childAttr.name)
 		}
-		ssn.RecordPodGroupEvent(job.PodGroup, v1.EventTypeNormal, string(scheduling.PodGroupUnschedulableType), "queue resource quota insufficient")
-		return util.Reject
-	})
+	}
 
-	// Register event handlers.
-	ssn.AddEventHandler(&framework.EventHandler{
-		AllocateFunc: func(event *framework.Event) {
-			job := ssn.Jobs[event.Task.Job]
-			attr := cp.queueOpts[job.Queue]
-			attr.allocated.Add(event.Task.Resreq)
-			metrics.UpdateQueueAllocated(attr.name, attr.allocated.MilliCPU, attr.allocated.Memory)
+	if attr.name == cp.rootQueue {
+		attr.guarantee = totalGuarantee
+		cp.totalGuarantee = totalGuarantee
+		attr.realCapability = cp.totalResource
+		attr.deserved = cp.totalResource
+	}
 
-			cp.updateShare(attr)
+	for _, childAttr := range attr.children {
+		realCapability := attr.realCapability.Clone().Sub(totalGuarantee).Add(childAttr.guarantee)
+		if childAttr.capability == nil {
+			childAttr.realCapability = realCapability
+		} else {
+			realCapability.MinDimensionResource(childAttr.capability, api.Infinity)
+			childAttr.realCapability = realCapability
+		}
+		oldDeserved := childAttr.deserved.Clone()
+		childAttr.deserved.MinDimensionResource(childAttr.realCapability, api.Infinity)
+		childAttr.deserved.MinDimensionResource(childAttr.request, api.Infinity)
 
-			klog.V(4).Infof("Capacity AllocateFunc: task <%v/%v>, resreq <%v>,  share <%v>",
-				event.Task.Namespace, event.Task.Name, event.Task.Resreq, attr.share)
-		},
-		DeallocateFunc: func(event *framework.Event) {
-			job := ssn.Jobs[event.Task.Job]
-			attr := cp.queueOpts[job.Queue]
-			attr.allocated.Sub(event.Task.Resreq)
-			metrics.UpdateQueueAllocated(attr.name, attr.allocated.MilliCPU, attr.allocated.Memory)
+		childAttr.deserved = helpers.Max(childAttr.deserved, childAttr.guarantee)
+		totalDeserved.Sub(oldDeserved).Add(childAttr.deserved)
+	}
 
-			cp.updateShare(attr)
+	// Check if the parent queue's deserved resources are less than the total deserved resources of child queues
+	if attr.deserved.LessPartly(totalDeserved, api.Zero) {
```

Comment:
`openSession` will not retuen err when there is an err of a plugin , so even thougt we return an err here, the `openSession` will not end and the plugin will contiune its logic, so does this check really take effect?

Reply:
In the latest implementation, if an error occurs during openSession, the injected functions will directly return the logic related to rejecting the scheduling. For example, in ssn.AddJobEnqueueableFn, it immediately returns a rejection, preventing the job from being enqueued. What do you think about this approach?